gcs/upload.py

```diff
@@ -280,6 +282,135 @@ def init_write_object_grpc(cls, db, request_iterator, context):
+                    request.write_object_spec.resource.bucket, context
+                ).metadata
+                upload = cls.__init_first_write_grpc(request, bucket, context)
+            elif upload is None:
+                return testbench.error.invalid(
+                    "Upload missing a first_message field", context
+                )
+
+            if request.HasField("object_checksums"):
+                # The object checksums may appear only in the first message *or* the last message, but not both
+                if first_message is None and request.finish_write == False:
+                    return testbench.error.invalid(
+                        "Object checksums can be included only in the first or last message",
+                        context,
+                    )
+                if object_checksums is not None:
+                    return testbench.error.invalid(
+                        "Duplicate object checksums in upload",
+                        context,
+                    )
+                object_checksums = request.object_checksums
+
+            data = request.WhichOneof("data")
+            if data == "checksummed_data":
+                checksummed_data = request.checksummed_data
+            elif data is None and request.finish_write:
+                # Handles final message with no data to insert.
+                upload.complete = True
+                continue
+
+            content = checksummed_data.content
+            crc32c_hash = (
+                checksummed_data.crc32c if checksummed_data.HasField("crc32c") else None
+            )
+            if crc32c_hash is not None:
+                actual_crc32c = crc32c.crc32c(content)
+                if actual_crc32c != crc32c_hash:
+                    return testbench.error.mismatch(
+                        "crc32c in checksummed data",
+                        crc32c_hash,
+                        actual_crc32c,
+                        context,
+                    )
+
+            # The testbench should ignore any request bytes that have already been persisted,
+            # thus we validate write_offset against persisted_size.
+            # https://github.com/googleapis/googleapis/blob/15b48f9ed0ae8b034e753c6895eb045f436e257c/google/storage/v2/storage.proto#L320-L329
+            if request.flush or should_checkpoint_session:
+                if request.write_offset < len(upload.media):
+                    range_start = len(upload.media) - request.write_offset
+                    content = testbench.common.partial_media(
+                        content, range_end=len(content), range_start=range_start
+                    )
+                upload.media += content
+            if request.finish_write:
+                upload.complete = True
+            if request.state_lookup:
+                yield storage_pb2.BidiWriteObjectResponse(
+                    persisted_size=len(upload.media)
+                )
+
+        if upload is None:
```

Comment:
Checking this after doing `upload.media += content` seems like a bug.

Seems like this check should be around line 350 or so?

Reply:
This is to handle cases with missing BidiWriteObjectRequest. I probably went overboard, and there's also a test case for this `gcs.upload.Upload.process_bidi_write_object_grpc(db, [], context)`
https://github.com/googleapis/storage-testbench/blob/65105b6fdf5efcf590326bac65650076b3677824/tests/test_upload.py#L646-L653

Reply:
Right, but if `upload` is None then line 359 would have raised an exception before you got here, right?   You probably want to detect the problem before the first use of `upload`.  Did I miss something?  If so, please just merge, otherwise please reorder the test to prevent ~crashes~ exceptions.

Reply:
Inside the request loop on L307, we're checking if `upload` is None before the first use. 

This extra check is outside the request loop. It covers the rare case of missing requests, and prevents crashes in the attempt to finalize an upload